lib/model/message.dart

```diff
@@ -435,4 +488,414 @@ class MessageStoreImpl extends PerAccountStoreBase with MessageStore {
+    }
+
+    outboxMessage._state = newState;
+    for (final view in _messageListViews) {
+      if (oldState == OutboxMessageState.hidden) {
+        view.addOutboxMessage(outboxMessage);
+      } else {
+        view.notifyListenersIfOutboxMessagePresent(localMessageId);
+      }
+    }
+  }
+
+  /// Send a message and create an entry of [OutboxMessage].
+  Future<void> _outboxSendMessage({
+    required MessageDestination destination,
+    required String content,
+    required String? realmEmptyTopicDisplayName,
+  }) async {
+    assert(!_disposed);
+    final localMessageId = _nextLocalMessageId++;
+    assert(!outboxMessages.containsKey(localMessageId));
+
+    final conversation = switch (destination) {
+      StreamDestination(:final streamId, :final topic) =>
+        StreamConversation(
+          streamId,
+          _processTopicLikeServer(
+            topic, realmEmptyTopicDisplayName: realmEmptyTopicDisplayName),
+          displayRecipient: null),
+      DmDestination(:final userIds) => DmConversation(allRecipientIds: userIds),
+    };
+
+    _outboxMessages[localMessageId] = OutboxMessage.fromConversation(
+      conversation,
+      localMessageId: localMessageId,
+      selfUserId: selfUserId,
+      timestamp: ZulipBinding.instance.utcNow().millisecondsSinceEpoch ~/ 1000,
+      contentMarkdown: content);
+
+    _outboxMessageDebounceTimers[localMessageId] = Timer(
+      kLocalEchoDebounceDuration,
+      () => _handleOutboxDebounce(localMessageId));
+
+    _outboxMessageWaitPeriodTimers[localMessageId] = Timer(
+      kSendMessageOfferRestoreWaitPeriod,
+      () => _handleOutboxWaitPeriodExpired(localMessageId));
+
+    try {
+      await _apiSendMessage(connection,
+        destination: destination,
+        content: content,
+        readBySender: true,
+        queueId: queueId,
+        localId: localMessageId.toString());
+    } catch (e) {
+      if (_disposed) return;
+      if (!_outboxMessages.containsKey(localMessageId)) {
+        // The message event already arrived; the failure is probably due to
+        // networking issues. Don't rethrow; the send succeeded
+        // (we got the event) so we don't want to show an error dialog.
+        return;
+      }
+      _outboxMessageDebounceTimers.remove(localMessageId)?.cancel();
+      _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
+      _updateOutboxMessage(localMessageId, newState: OutboxMessageState.failed);
+      rethrow;
+    }
+    if (_disposed) return;
+    if (!_outboxMessages.containsKey(localMessageId)) {
+      // The message event already arrived; nothing to do.
+      return;
+    }
+    // The send request succeeded, so the message was definitely sent.
+    // Cancel the timer that would have had us start presuming that the
+    // send might have failed.
+    _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
+    if (_outboxMessages[localMessageId]!.state
+          == OutboxMessageState.waitPeriodExpired) {
```

Comment:
Neat, glad that change turned out to be clean to make! (Following up on https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2101148783.)

The diffstat is:
 lib/model/message.dart       | 28 ++++++++++++++++++++--------
 test/model/message_test.dart | 28 ++++++++++++++++++++++++++++
 2 files changed, 48 insertions(+), 8 deletions(-)

and outside the new test, most of that is comments; I count 6+/1- other than tests and comments.